cumulusci/utils/clean_invalid_references.py

```diff
@@ -0,0 +1,365 @@
+USER_PERMISSION = PermissionElementXPath(".//userPermissions", "name")
+TAB_PERMISSION = PermissionElementXPath(".//tabVisibilities", "tab")
+APP_PERMISSION = PermissionElementXPath(".//applicationVisibilities", "application")
+APEXCLASS_PERMISSION = PermissionElementXPath(".//classAccesses", "apexClass")
+APEXPAGE_PERMISSION = PermissionElementXPath(".//pageAccesses", "apexPage")
+FLOW_PERMISSION = PermissionElementXPath(".//flowAccesses", "flow")
+CUSTOMPERM_PERMISSION = PermissionElementXPath(".//customPermissions", "name")
+CUSTOMSETTING_PERMISSION = PermissionElementXPath(".//customSettingAccesses", "name")
+CUSTOMMETADATA_PERMISSION = PermissionElementXPath(
+    ".//customMetadataTypeAccesses", "name"
+)
+
+PACKAGEXML_DICT = {
+    # PERMISSION : is_child
+    "CustomObject": {
+        OBJECT_PERMISSION: False,
+        CUSTOMSETTING_PERMISSION: False,
+        CUSTOMMETADATA_PERMISSION: False,
+        RECORDTYPE_PERMISSION: True,
+        FIELD_PERMISSION: True,
+    },
+    "ApexClass": APEXCLASS_PERMISSION,
+    "ApexPage": APEXPAGE_PERMISSION,
+    "Flow": FLOW_PERMISSION,
+    "CustomTab": TAB_PERMISSION,
+    "CustomApplication": APP_PERMISSION,
+    "CustomPermission": CUSTOMPERM_PERMISSION,
+}
+
+FOLDER_PERM_DICT = {
+    "objects": [OBJECT_PERMISSION, CUSTOMSETTING_PERMISSION, CUSTOMMETADATA_PERMISSION],
+    "fields": [FIELD_PERMISSION],
+    "tabs": [TAB_PERMISSION],
+    "applications": [APP_PERMISSION],
+    "customPermissions": [CUSTOMPERM_PERMISSION],
+    "pages": [APEXPAGE_PERMISSION],
+    "classes": [APEXCLASS_PERMISSION],
+    "flows": [FLOW_PERMISSION],
+    "recordTypes": [RECORDTYPE_PERMISSION],
+    "userPermissions": [USER_PERMISSION],
+}
+
+
+def run_queries_in_parallel(
+    queries: Dict[str, str], run_query: Callable[[str], dict], num_threads: int = 4
+):
+    """Accepts a set of queries structured as {'query_name': 'query'}
+    and a run_query function that runs a particular query. Runs queries in parallel and returns the queries"""
+    results_dict = {}
+
+    with ThreadPoolExecutor(max_workers=num_threads) as executor:
+        futures = {
+            query_name: executor.submit(run_query, query)
+            for query_name, query in queries.items()
+        }
+
+    for query_name, future in futures.items():
+        try:
+            query_result = future.result()
+            results_dict[query_name] = query_result
+        except Exception as e:
+            raise Exception(f"Error executing query '{query_name}': {type(e)}: {e}")
+        else:
+            queries.pop(query_name, None)
+
+    return results_dict
+
+
+def entities_from_query(sf: Salesforce):
+    # Define the query
+    def run_query(path: str):
+        urlpath = sf.base_url + path
+        method = "GET"
+        return sf._call_salesforce(method, urlpath).json()
+
+    # Queries
+    queries = {}
+    queries["userPermissions"] = "sobjects/PermissionSet/describe"
```

Comment:
I like running these queries using the sObjects rather than the Tooling API, which was what I originally had in mind. I would guess this approach is faster; it's definitely more elegant.

Reply:
One quibble: most of these aren't _really_ queries, in the SOQL sense - they're API calls. Maybe adjust the naming to clarify? I'm also fine if you want to keep it like this, though.

Reply:
yep I will change the variable name